File: AIReader/AIReader/Domain/Books/ReadingPlace.swift
import Foundation

/// Where a reader is in a book, in terms that mean the same on every device:
/// the chapter, how far into it, and the words at that point. Offsets differ
/// between apps — UTF-16 units here, bytes on the Kindle — and between
/// renderings, so a place is found again by its words, and by its fraction
/// when the words cannot be found.
///
/// The words are compared loosely: every run of whitespace, breaking or not,
/// counts as one space, and illustration placeholders do not count at all,
/// since the two apps render those differently.
struct ReadingPlace: Equatable, Sendable, Codable {
    var chapter: Int
    /// How far into the chapter's text, from 0 to 1.
    var fraction: Double
    /// The text starting at the place, a few words of it, normalized.
    var snippet: String

    /// How much text is kept as the snippet, in UTF-16 units, before it is
    /// cut back to a word boundary.
    static let snippetLength = 80

    /// The place at `offset` (UTF-16) of a chapter's text.
    init(chapter: Int, chapterText: String, offset: Int) {
        let text = chapterText as NSString
        let at = min(max(offset, 0), text.length)
        self.chapter = chapter
        self.fraction = text.length > 0 ? Double(at) / Double(text.length) : 0

        let ahead = text.substring(with: NSRange(location: at, length: min(text.length - at, Self.snippetLength * 3)))
        let normalized = Self.normalize(ahead).text as NSString
        var end = min(normalized.length, Self.snippetLength)
        // Cut back to a space so the snippet is whole words, unless that
        // would leave nothing.
        if end < normalized.length {
            var cut = end
            while cut > 0, normalized.character(at: cut - 1) != Self.space { cut -= 1 }
            if cut > 0 { end = cut }
        }
        self.snippet = normalized.substring(to: end).trimmingCharacters(in: .whitespaces)
    }

    init(chapter: Int, fraction: Double, snippet: String) {
        self.chapter = chapter
        self.fraction = fraction
        self.snippet = snippet
    }

    /// The UTF-16 offset in the chapter's text this place stands for: where
    /// the snippet is found — the nearest of its occurrences to the fraction
    /// — else the fraction alone.
    func resolve(in chapterText: String) -> Int {
        let original = chapterText as NSString
        let guess = Double(original.length) * min(max(fraction, 0), 1)
        guard !snippet.isEmpty else { return Int(guess.rounded(.down)) }

        let (normalized, map) = Self.normalize(chapterText)
        let haystack = normalized as NSString
        let expected = Double(haystack.length) * min(max(fraction, 0), 1)
        var best: Int?
        var from = 0
        while from < haystack.length {
            let found = haystack.range(of: snippet, range: NSRange(location: from, length: haystack.length - from))
            guard found.location != NSNotFound else { break }
            if best.map({ abs(Double(found.location) - expected) < abs(Double($0) - expected) }) ?? true {
                best = found.location
            }
            from = found.location + 1
        }
        guard let best, best < map.count else { return Int(guess.rounded(.down)) }
        return map[best]
    }

    // MARK: - Loose comparison

    private static let space: unichar = 0x20
    private static let placeholder: unichar = 0xFFFC

    /// `text` with whitespace collapsed and placeholders dropped, and for
    /// each unit of the result the offset it came from.
    static func normalize(_ text: String) -> (text: String, map: [Int]) {
        let source = text as NSString
        var units: [unichar] = []
        var map: [Int] = []
        var pendingSpace = false
        for index in 0..<source.length {
            let c = source.character(at: index)
            if c == placeholder { continue }
            if UnicodeScalar(c).map({ CharacterSet.whitespacesAndNewlines.contains($0) }) ?? false {
                pendingSpace = !units.isEmpty
                continue
            }
            if pendingSpace {
                units.append(space)
                map.append(index)
                pendingSpace = false
            }
            units.append(c)
            map.append(index)
        }
        return (String(utf16CodeUnits: units, count: units.count), map)
    }
}

File: AIReaderKindle/src/Domain/Books/ReadingPlace.cpp
#include "ReadingPlace.hpp"

#include "../../Support/Text.hpp"

#include <glib.h>

#include <algorithm>
#include <cmath>
#include <cstdlib>

namespace {

/// `text` with whitespace collapsed and illustration placeholders dropped,
/// and for each byte of the result the byte offset it came from.
struct Normalized {
    std::string text;
    std::vector<int> map;
};

Normalized normalize(const std::string& text) {
    Normalized out;
    bool pendingSpace = false;
    const char* base = text.c_str();
    const char* end = base + text.size();
    for (const char* p = base; p < end;) {
        gunichar c = g_utf8_get_char(p);
        const char* next = g_utf8_next_char(p);
        if (c == 0xFFFC) { p = next; continue; }
        if (g_unichar_isspace(c)) {
            pendingSpace = !out.text.empty();
            p = next;
            continue;
        }
        if (pendingSpace) {
            out.text += ' ';
            out.map.push_back(static_cast<int>(p - base));
            pendingSpace = false;
        }
        for (const char* q = p; q < next; ++q) {
            out.text += *q;
            out.map.push_back(static_cast<int>(p - base));
        }
        p = next;
    }
    return out;
}

}  // namespace

ReadingPlace ReadingPlace::at(int chapter, const std::string& chapterText, int offset) {
    int size = static_cast<int>(chapterText.size());
    int start = std::min(std::max(offset, 0), size);
    ReadingPlace place;
    place.chapter = chapter;
    place.fraction = size > 0 ? static_cast<double>(start) / size : 0;

    std::string ahead = chapterText.substr(start, std::min(size - start, snippetLength * 3));
    std::string normalized = normalize(ahead).text;
    int end = std::min(static_cast<int>(normalized.size()), snippetLength);
    // Back off a cut inside a character, then to a space so the snippet is
    // whole words, unless that would leave nothing.
    while (end > 0 && end < static_cast<int>(normalized.size()) && (static_cast<unsigned char>(normalized[end]) & 0xC0) == 0x80) --end;
    if (end < static_cast<int>(normalized.size())) {
        int cut = end;
        while (cut > 0 && normalized[cut - 1] != ' ') --cut;
        if (cut > 0) end = cut;
    }
    place.snippet = Text::trim(normalized.substr(0, end));
    return place;
}

int ReadingPlace::resolve(const std::string& chapterText) const {
    double clamped = std::min(std::max(fraction, 0.0), 1.0);
    auto byFraction = [&] {
        int offset = static_cast<int>(std::floor(chapterText.size() * clamped));
        // Never land inside a character.
        while (offset > 0 && offset < static_cast<int>(chapterText.size())
               && (static_cast<unsigned char>(chapterText[offset]) & 0xC0) == 0x80) --offset;
        return offset;
    };
    if (snippet.empty()) return byFraction();

    Normalized normalized = normalize(chapterText);
    double expected = normalized.text.size() * clamped;
    // The nearest occurrence to where the fraction says it should be.
    std::optional<size_t> best;
    for (size_t found = normalized.text.find(snippet); found != std::string::npos; found = normalized.text.find(snippet, found + 1)) {
        if (!best || std::fabs(static_cast<double>(found) - expected) < std::fabs(static_cast<double>(*best) - expected)) best = found;
    }
    if (!best || *best >= normalized.map.size()) return byFraction();
    return normalized.map[*best];
}
